doc/comp.1
.\" Process this file with
.\" groff -man -Tutf8 comp.1
.\"
.TH COMP 1 2018-01-25 comp
.SH NAME
comp \- Curses Omni Media Player
.SH SYNOPSIS
\fBcomp\fR [\fB-h\fR] [\fB-v\fR] [\fB-e\fR {json,mpv,youtube-dl}]
[\fB-c \fICONFIG\fR] [\fB--vid \fIVID\fR] [\fB--vo \fIDRIVER\fR]
[\fB-f \fIYTDL_FORMAT\fR] \fIplaylist\fR
.SH DESCRIPTION
\fBcomp\fR is a 
.BR mpv (1)
front-end using
.BR curses (3).
It has basic media player functions and can to extract playlists from multiple
sources such as media sites supported by
.BR youtube-dl (1),
local and direct URL to video/audio and its own JSON playlist format.
.SH OPTIONS
.SS Positional arguments
.TP
.B playlist
path or URL to the playlist
.SS Optional arguments
.TP 
.B -h, --help
show this help message and exit
.TP
.B -v, --version
show program's version number and exit
.TP
.B -e \fR{json,mpv,youtube-dl}, \fB--extractor \fR{json,mpv,youtube-dl}
playlist extractor, default is \fIyoutube-dl
.TP
.B -c \fICONFIG, \fB--config \fICONFIG
path to the configuration file
.TP 
.B --vid \fIVID
initial video channel. \fIauto\fR selects the default, \fIno\fR disables video
.TP
.B --vo \fIDRIVER
specify the video output backend to be used. See 
.I VIDEO OUTPUT DRIVERS
in
.BR mpv (1)
for details and descriptions of available drivers
.TP 
.B -f \fIYTDL_FORMAT\fR, \fB--format \fIYTDL_FORMAT
video format/quality to be passed to youtube-dl
.SH KEYBOARD CONTROL
.SS Bindings inherited from mpv
For convenience purpose, I try to mimic 
.BR mpv (1)
default keybindings, but many are slightly different from
.BR mpv (1)
exact behaviour (mainly because of the lack of keys which are unsupported by
.BR curses (3)).
So I will list all of them here for you to compare:
.TP
.B Left and Right
Seek backward/forward 5 seconds. Shifted arrow does a 1 second seek.
.TP
.B Up and Down
Seek backward/forward 1 minute.
.TP
.B [ and ]
Decrease/increase current playback speed by 10%.
.TP
.B { and }
Halve/double current playback speed.
.TP
.B Backspace
Reset playback speed to normal.
.TP
.B < and >
Go backward/forward in the playlist.
.TP
.B Return
Start playing.
.TP
.B Space / p
Pause (pressing again unpauses).
.TP
.B .
Step forward. Pressing once will pause, every consecutive press will play
one frame and then go into pause mode again.
.TP
.B ,
Step backward. Pressing once will pause, every consecutive press will play
one frame in reverse and then go into pause mode again.
.TP
.B q
Stop playing and quit.
.TP
.B / and *
Decrease/increase volume.
.TP
.B 9 and 0
Decrease/increase volume.
.TP
.B m
Mute sound.
.TP
.B _
Cycle through the available video tracks.
.TP
.B #
Cycle through the available audio tracks.
.TP
.B f
Toggle fullscreen.
.TP
.B T
Toggle stay-on-top.
.TP
.B w and e
Decrease/increase pan-and-scan range.
.TP
.B o or P
Show progression bar, elapsed time and total duration on the OSD.
.TP
.B O
Toggle OSD states between normal and playback time/duration.
.TP
.B v
Toggle subtitle visibility.
.TP
.B j and J
Cycle through the available subtitles.
.TP
.B x and z
Adjust subtitle delay by +/- 0.1 seconds.
.TP
.B l
Set/clear A-B loop points.
.TP
.B L
Toggle infinite looping.
.TP
.B Ctrl-+ and Ctrl--
Adjust audio delay (A/V sync) by +/- 0.1 seconds.
.TP
.B u
Switch between applying no style overrides to SSA/ASS subtitles, and
overriding them almost completely with the normal subtitle style.
.TP
.B V
Toggle subtitle VSFilter aspect compatibility mode.
.TP
.B r and t
Move subtitles up/down.
.TP
.B s
Take a screenshot.
.TP
.B S
Take a screenshot, without subtitles.
.TP
.B Alt-s
Take a screenshot each frame.
.TP
.B Page Up and Page Down
Seek to the beginning of the previous/next chapter.
.TP
.B d
Activate/deactivate deinterlacer.
.TP
.B A
Cycle aspect ratio override.
.TP
.B 1 and 2
Adjust contrast.
.TP
.B 3 and 4
Adjust brightness.
.TP
.B 5 and 6
Adjust gamma.
.TP
.B 7 and 8
Adjust saturation.
.TP
.B Alt-0
Resize video window to half its original size.
.TP
.B Alt-1
Resize video window to its original size.
.TP
.B Alt-2
Resize video window to double its original size.
.TP
.B E
Cycle through editions.
.SS Movements and selections
The following keybindings are Emacs-like since most characters are taken by
.BR mpv (1).
.TP
.B Ctrl-p and Ctrl-n
Move a single line up/down.
.TP
.B Alt-v and Ctrl-v
Move a single page up/down.
.TP
.B Ctrl-< and Ctrl->
Move to the beginning/end of the playlist.
.TP
.B Home and End
Move to the beginning/end of the playlist.
.TP
.B Ctrl-Space
Deselect/reselect the current entry and move down a line.
.SS Playlist manipulation
.TP
.B Ctrl-o
Open playlist.
.TP
.B Ctrl-i
Insert playlist.
.TP
.B Ctrl-f and Alt-f
Search forward/backward for a pattern.
.TP
.B Alt-m
Cycle through playing modes.
.TP
.B Delete
Delete the current entry.
.TP
.B W
Save the current playlist under JSON format.
.TP
.B F5
Redraw the screen content.
.TP
.B :
Execute a mpv command.
.SH FILES
.TP
.I ~/.config/comp/settings.ini
per user configuration file
.SH EXAMPLES
.nf R
Open a JSON playlist:
.ft B
	comp -e json test/playlist.json

.ft R
Open a Youtube playlist with video height lower than 720:
.ft B
	comp -f '[height<720]' https://www.youtube.com/watch?list=PLnk14Iku8QM7R3ARnrj1TwYSZleF-i7jT
.SH BUGS
.PP
Media durations are not extracted from online playlists as youtube-dl
extract_flat option is used. This is rather a feature to save up bandwidth than
a bug because a track's duration is updated when it's played.
.SH AUTHOR
Written by Nguyễn Gia Phong.
.SH "SEE ALSO"
.BR mpv (1),
.BR youtube-dl (1)


comp.py
#!/usr/bin/env python3
 
# comp - Curses Online Media Player
# Copyright (C) 2017  Raphael McSinyx
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import curses
import json
from argparse import ArgumentParser
from configparser import ConfigParser
from itertools import cycle
from os.path import expanduser
from random import choice
from time import gmtime, strftime
from threading import Thread

from mpv import MPV

MODES = ('play-current', 'play-all', 'play-selected', 'repeat-current',
         'repeat-all', 'repeat-selected', 'shuffle-all', 'shuffle-selected')


def setno(data, keys):
    """Set all keys of each track in data to False."""
    for key in keys:
        for track in data:
            track[key] = False


def playlist(mode):
    """Return a generator of tracks to be played."""
    action, choose_from = mode.split('-')
    if choose_from == 'all': tracks = data
    else: tracks = [track for track in data if track[choose_from]]
    # Somehow yield have to be used instead of returning a generator
    if action == 'play':
        for track in tracks: yield track
    elif action == 'repeat':
        for track in cycle(tracks): yield track
    elif tracks:
        while True: yield choice(tracks)


def play():
    for track in playlist(mode):
        setno(data, ['playing'])
        data[data.index(track)]['playing'] = True
        reprint(stdscr, data[start : start+curses.LINES-3])
        mp.play('https://youtu.be/' + track['url'])
        mp.wait_for_playback()


def secpair2hhmmss(pos, duration):
    """Quick hack to convert a pair of seconds to HHMMSS / HHMMSS
    string as MPV.get_property_osd_string isn't available.
    """
    if pos is None: return ''
    postime, durationtime = gmtime(pos), gmtime(duration)
    # Let's hope media durations are shorter than a day
    timestr = '%M:%S' if duration < 3600 else '%H:%M:%S'
    return '{} / {}'.format(strftime(timestr, postime),
                            strftime(timestr, durationtime))


def update_status_line(stdscr, mp):
    left = ' ' + secpair2hhmmss(mp._get_property('time-pos', int),
                                mp._get_property('duration', int))
    right = ' {} {}{} '.format(mode,
                               ' ' if mp._get_property('mute', bool) else 'A',
                               ' ' if mp._get_property('vid') == 'no' else 'V')
    if left != ' ':
        left += ' | ' if mp._get_property('pause', bool) else ' > '
        stdscr.addstr(curses.LINES - 2, 0, left, curses.color_pair(8))
        title_len = curses.COLS - len(left + right)
        center = mp._get_property('media-title').ljust(title_len)[:title_len]
        stdscr.addstr(curses.LINES - 2, len(left), center,
                      curses.color_pair(8) | curses.A_BOLD)
        stdscr.addstr(curses.LINES - 2, len(left + center), right,
                      curses.color_pair(8))
    else:
        stdscr.addstr(curses.LINES - 2, 0, right.rjust(curses.COLS),
                      curses.color_pair(8))
    stdscr.refresh()


def reattr(stdscr, y, track):
    invert = 8 if track['current'] else 0
    if track['error']:
        stdscr.chgat(y, 0, curses.color_pair(1 + invert) | curses.A_BOLD)
    elif track['playing']:
        stdscr.chgat(y, 0, curses.color_pair(3 + invert) | curses.A_BOLD)
    elif track['selected']:
        stdscr.chgat(y, 0, curses.color_pair(5 + invert) | curses.A_BOLD)
    elif invert:
        stdscr.chgat(y, 0, curses.color_pair(12) | curses.A_BOLD)
    else:
        stdscr.chgat(y, 0, curses.color_pair(0) | curses.A_NORMAL)


def reprint(stdscr, data2print):
    stdscr.clear()
    stdscr.addstr(0, curses.COLS-12, 'URL')
    stdscr.addstr(0, 1, 'Title')
    stdscr.chgat(0, 0, curses.color_pair(10) | curses.A_BOLD)
    for i, track in enumerate(data2print):
        y = i + 1
        stdscr.addstr(y, 0, track['url'].rjust(curses.COLS - 1))
        stdscr.addstr(y, 1, track['title'][:curses.COLS-14])
        reattr(stdscr, y, track)
    update_status_line(stdscr, mp)


def move(stdscr, data, y, delta):
    global start
    if start + y + delta < 1:
        if start + y == 1:
            return 1
        start = 0
        setno(data, ['current'])
        data[0]['current'] = True
        reprint(stdscr, data[:curses.LINES-3])
        return 1
    elif start + y + delta > len(data):
        if start + y == len(data):
            return curses.LINES - 3
        start = len(data) - curses.LINES + 3
        y = curses.LINES - 3
        setno(data, ['current'])
        data[-1]['current'] = True
        reprint(stdscr, data[-curses.LINES+3:])
        return y

    if y + delta < 1:
        start += y + delta - 1
        y = 1
        setno(data, ['current'])
        data[start]['current'] = True
        reprint(stdscr, data[start : start+curses.LINES-3])
    elif y + delta > curses.LINES - 3:
        start += y + delta - curses.LINES + 3
        y = curses.LINES - 3
        setno(data, ['current'])
        data[start + curses.LINES - 4]['current'] = True
        reprint(stdscr, data[start : start+curses.LINES-3])
    else:
        data[start + y - 1]['current'] = False
        reattr(stdscr, y, data[start + y - 1])
        y = y + delta
        data[start + y - 1]['current'] = True
        reattr(stdscr, y, data[start + y - 1])
        stdscr.refresh()
    return y


parser = ArgumentParser(description="console/curses online media mp")
parser.add_argument('-j', '--json-playlist', required=False,
                    help='path to playlist in JSON format')
args = parser.parse_args()

config = ConfigParser()
config.read(expanduser('~/.config/comp/settings.ini'))
mode = config.get('comp', 'play-mode', fallback='play-all')
video = config.get('mpv', 'video', fallback='auto')
video_output = config.get('mpv', 'video-output', fallback='')
ytdlf = config.get('youtube-dl', 'format', fallback='best')

with open(args.json_playlist) as f:
    data = json.load(f)
setno(data, ['error', 'playing', 'selected', 'current'])

stdscr = curses.initscr()
curses.noecho()
curses.cbreak()
stdscr.keypad(True)
curses.curs_set(False)
curses.start_color()
curses.use_default_colors()
curses.init_pair(1, 1, -1)
curses.init_pair(2, 2, -1)
curses.init_pair(3, 3, -1)
curses.init_pair(4, 4, -1)
curses.init_pair(5, 5, -1)
curses.init_pair(6, 6, -1)
curses.init_pair(7, 7, -1)
curses.init_pair(8, -1, 7)
curses.init_pair(9, -1, 1)
curses.init_pair(10, -1, 2)
curses.init_pair(11, -1, 3)
curses.init_pair(12, -1, 4)
curses.init_pair(13, -1, 5)
curses.init_pair(14, -1, 6)

mp = MPV(input_default_bindings=True, input_vo_keyboard=True,
         ytdl=True, ytdl_format=ytdlf)
if video_output: mp['vo'] = video_output
mp._set_property('vid', video)
mp.observe_property('mute', lambda foo: update_status_line(stdscr, mp))
mp.observe_property('pause', lambda foo: update_status_line(stdscr, mp))
mp.observe_property('time-pos', lambda foo: update_status_line(stdscr, mp))
mp.observe_property('vid', lambda foo: update_status_line(stdscr, mp))

# Print initial content
start = 0
y = 1
data[0]['current'] = True
reprint(stdscr, data[:curses.LINES-3])

# mpv keys: []{}<>.,qQ/*90m-#fTweoPOvjJxzlLVrtsSIdA
# yuighkcbn
c = stdscr.getch()
while c != 113:     # letter q
    if c == curses.KEY_RESIZE:
        curses.update_lines_cols()
        start += y - 1
        y = 1
        reprint(stdscr, data[start : start+curses.LINES-3])
    elif c in (106, curses.KEY_DOWN):   # letter j or down arrow
        y = move(stdscr, data, y, 1)
    elif c in (107, curses.KEY_UP):     # letter k or up arrow
        y = move(stdscr, data, y, -1)
    elif c == curses.KEY_PPAGE:     # page up
        y = move(stdscr, data, y, 4 - curses.LINES)
    elif c == curses.KEY_NPAGE:     # page down
        y = move(stdscr, data, y, curses.LINES - 4)
    elif c == curses.KEY_HOME:  # home
        y = move(stdscr, data, y, -len(data))
    elif c == curses.KEY_END:   # end
        y = move(stdscr, data, y, len(data))
    elif c == 109:  # letter m
        mode = MODES[(MODES.index(mode) + 1) % 8]
        update_status_line(stdscr, mp)
    elif c == 77:   # letter M
        mode = MODES[(MODES.index(mode) - 1) % 8]
        update_status_line(stdscr, mp)
    elif c == 112:  # letter p
        mp._set_property('pause', False, bool)
        play_thread = Thread(target=play)
        play_thread.daemon = True
        play_thread.start()
    elif c == 32:   # space
        mp._set_property('pause', not mp._get_property('pause', bool), bool)
    elif c == 99:   # letter c
        data[start + y - 1]['selected'] = not data[start + y - 1]['selected']
        y = move(stdscr, data, y, 1)
    elif c == 97:   # letter a
        mp._set_property('mute', not mp._get_property('mute', bool), bool)
    elif c == 118:  # letter v
        mp._set_property('vid', 'auto' if mp._get_property('vid') == 'no' else 'no')
    c = stdscr.getch()

curses.nocbreak()
stdscr.keypad(False)
curses.echo()
curses.endwin()

del mp
